### DISCUSSION TITLE
next js prisma - undefined when in vercel deployment

### DISCUSSION THREAD

**Discussion by sunrunner4kr-simon** [prisma/prisma#6651] [Q&A] [answered]
Hi all

So...I have an api in Next js that uses Prisma Client. Prisma is imported from the global object defined in prisma.ts

Locally everything builds and runs fine. I get no errors and the prisma variable is defined. However, when it's deployed in Vercel, prisma is undefined...I can't work out why.

It was also working fine before when I was calling a prisma instance in each page - I changed it to declare prisma globally (so I wouldn't get multiple instances)

If anyone has any suggestions, I'd much appreciate it.

```js
import eBayApi from "@hendt/ebay-api";
import prisma from "../../lib/prisma.ts";

const eBay = new eBayApi({});

export default async (req, res) => {
  // Access the provided 'page' and 'limt' query parameters
  const code = req.query.code; // this is provided from eBay
  console.log(code);

  try {
    const token = await eBay.OAuth2.getToken(code);

    const env_variable = await prisma.env_variable.upsert({
      where: {
        variable: "EBAY_TOKEN",
      },
      update: { value: token },
      create: {
        variable: "EBAY_TOKEN",
        value: token,
      },
    });

    if (env_variable) {
      console.log("New Token Stored in DB");
    } else console.log("Failed to store new Token");

    res.status(200);
    res.writeHead(302, {
      Location: "/orders",
      //add other headers here...
    });
    res.end();
  } catch (e) {
    console.error(e);
    res.status(400).end();
  }

  res.writeHead(302, {
    Location: "/orders",
    //add other headers here...
  });
  res.end();
};
```

prisma.ts

```ts
import { PrismaClient } from "@prisma/client"

declare global {
    namespace NodeJS {
        interface Global {
            prisma: PrismaClient;
        }
    }
}

let prisma: PrismaClient;

if (process.env.NODE_ENV === "production") {
  prisma = new PrismaClient()
} else {
  if (!global.prisma) {
    global.prisma = new PrismaClient()
  }

  prisma = global.prisma
}

export default prisma

```
My Vercel error: -

```
2021-04-18T19:06:18.680Z 869eb228-423a-4d6a-b05a-f95f5e843c88 ERROR TypeError: Cannot read property 'upsert' of undefined at exports.modules.5712.webpack_exports.default (/var/task/nextjs-store/.next/server/pages/api/success.js:55:126) at processTicksAndRejections (internal/process/task_queues.js:93:5) at async apiResolver (/var/task/nextjs-store/node_modules/next/dist/next-server/server/api-utils.js:8:1) at async Server.handleApiRequest (/var/task/nextjs-store/node_modules/next/dist/next-server/server/next-server.js:67:462) at async Object.fn (/var/task/nextjs-store/node_modules/next/dist/next-server/server/next-server.js:59:492) at async Router.execute (/var/task/nextjs-store/node_modules/next/dist/next-server/server/router.js:25:67) at async Server.run (/var/task/nextjs-store/node_modules/next/dist/next-server/server/next-server.js:69:1042) at async Server.handleRequest (/var/task/nextjs-store/node_modules/next/dist/next-server/server/next-server.js:34:504) at async Server. (/var/task/nextjs-store/___next_launcher.js:26:9)

**Comment by 2color** [comment]
Hey @sunrunner4kr-simon,

Could you share your Prisma schema? I think the error might be related to the `env_variable` model. 

Are you getting this error when calling the function or during the build?

I would try adding a console.log where you instantiate Prisma (in `prisma.ts`) to see how Prisma is instantiated.

**Reply by sunrunner4kr-simon** [reply]
Thanks for helping!

It's when I call the function.

Here's the schema: - 

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mysql"
  url      = env("DATABASE_URL")
}

model order {
  orderId          String     @id
  lastModifiedDate String
  lineItems        lineItem[]
}

model lineItem {
  lineItemId String  @id
  sku        String
  order      order?  @relation(fields: [orderId], references: [orderId])
  orderId    String?

  @@index([orderId], name: "orderId")
}

model inventoryitem {
  sku                String @id
  productDescription String
  quantity           Int    @default(1)
  location           Int    @unique
}

model location {
  id      Int    @id
  barcode String
  state   String @default("EMPTY")
  sku     String @default("")
}

model Account {
  id                 Int       @id @default(autoincrement())
  compoundId         String    @unique @map(name: "compound_id")
  userId             Int       @map(name: "user_id")
  providerType       String    @map(name: "provider_type")
  providerId         String    @map(name: "provider_id")
  providerAccountId  String    @map(name: "provider_account_id")
  refreshToken       String?   @map(name: "refresh_token")
  accessToken        String?   @map(name: "access_token")
  accessTokenExpires DateTime? @map(name: "access_token_expires")
  createdAt          DateTime  @default(now()) @map(name: "created_at")
  updatedAt          DateTime  @default(now()) @map(name: "updated_at")


  @@index([providerAccountId], name: "providerAccountId")
  @@index([providerId], name: "providerId")
  @@index([userId], name: "userId")
  @@map(name: "accounts")
}

model Session {
  id           Int      @id @default(autoincrement())
  userId       Int      @map(name: "user_id")
  expires      DateTime
  sessionToken String   @unique @map(name: "session_token")
  accessToken  String   @unique @map(name: "access_token")
  createdAt    DateTime @default(now()) @map(name: "created_at")
  updatedAt    DateTime @default(now()) @map(name: "updated_at")

  @@map(name: "sessions")
}

model User {
  id            Int       @id @default(autoincrement())
  name          String?
  email         String?   @unique
  emailVerified DateTime? @map(name: "email_verified")
  image         String?
  createdAt     DateTime  @default(now()) @map(name: "created_at")
  updatedAt     DateTime  @default(now()) @map(name: "updated_at")
  username      String    @unique
  password      String
  salt          String

  @@map(name: "users")
}

model VerificationRequest {
  id         Int      @id @default(autoincrement())
  identifier String
  token      String   @unique
  expires    DateTime
  createdAt  DateTime @default(now()) @map(name: "created_at")
  updatedAt  DateTime @default(now()) @map(name: "updated_at")

  @@map(name: "verification_requests")
}

model env_variable {
  variable String @unique
  value    String
}
```

Will look at adding the logging

**Reply by sunrunner4kr-simon** [reply]
Not quite sure what I'm looking for in the prisma instantiate, but here's the console output of prisma

![image](https://user-images.githubusercontent.com/58920692/115392385-6b4abc80-a1d8-11eb-9108-65e13d9df841.png)

**Reply by sunrunner4kr-simon** [reply]
I change env_variable to Variable to match the prisma model naming convention. Still getting the same problem.

```prisma
model Variable {
  variable String @unique
  value    String
}
```

**Reply by 2color** [reply]
> Not quite sure what I'm looking for in the prisma instantiate, but here's the console output of prisma

Ok that already looks good. Can you also try to console.log the prisma instance within the request handler?

**Reply by 2color** [reply]
Another thing that might cause this is the fact that you're importing the `prisma` module with the `.ts` extension.

Try  updating the import as follows:
```ts
import prisma from "../../lib/prisma";
```

**Reply by sunrunner4kr-simon** [reply]
I had that originally...but it could never resolve it

![image](https://user-images.githubusercontent.com/58920692/115396044-79024100-a1dc-11eb-8d78-ca5dca1d400d.png)

**Reply by sunrunner4kr-simon** [reply]
And this is in the handler: - 

![image](https://user-images.githubusercontent.com/58920692/115396472-ed3ce480-a1dc-11eb-8bd1-eae4a0420722.png)

But these are the console logs from my local dev env that works...

**Reply by sunrunner4kr-simon** [reply]
This is the console logging in Vercel

![image](https://user-images.githubusercontent.com/58920692/115397549-232e9880-a1de-11eb-8ac5-bd32cc6f5300.png)

it dies before it reaches the handler logging...so it's failing on build, not running the function

**Reply by 2color** [reply]
Ok, this might be related to the fact that `/api/success.js` is a JavaScript file importing a TypeScript file that's somehow messing with the Webpack magic that Next.js does.

See error:
```
undefined at exports.modules.5712.webpack_exports.default
```

Can you maybe try migrating `/api/success.js` to TypeScript?

**Reply by sunrunner4kr-simon** [reply]
From my very limited understanding, I thought that next js should translate the ts to js when calling as "../../lib/prisma";

so it's working with js only..

but it seems to really struggle with that, i've no clue as to why, and the tutorial that I got the prisma next js global declaration from had it working as above.

So maybe there's something wrong with my implementation in handling that...but I haven't found anything searching. Probably because I don't know what to search for really.

**Reply by sunrunner4kr-simon** [reply]
This: -
https://www.prisma.io/docs/support/help-articles/nextjs-prisma-client-dev-practices

**Reply by 2color** [reply]
> From my very limited understanding, I thought that next js should translate the ts to js when calling as "../../lib/prisma";

It should, however, it's hard to say without looking at how your repository is set up. Perhaps you can create a reproduction repository so that we can take a look and resolve the issue you're experiencing. 

The example you shared is in TypeScript, but I would implement the same in JavaScript if that's what you're using for the rest of your project.

Try the following:
```js
import { PrismaClient } from '@prisma/client'

let prisma

// check to use this workaround only in development and not in production
if (process.env.NODE_ENV === 'production') {
  prisma = new PrismaClient()
} else {
  if (!global.prisma) {
    global.prisma = new PrismaClient()
  }

  prisma = global.prisma
}

export default prisma
```

**Reply by sunrunner4kr-simon** [reply]
Thanks

So...I'm now no longer getting problems resolving the prisma module.
So progress!!

Just tried it all in Vercel again, and still getting the error

2021-04-20T18:12:55.922Z	74782da2-8f0f-448c-bd44-ee81ead879ed	ERROR	TypeError: Cannot read property 'upsert' of undefined
    at exports.modules.5712.__webpack_exports__.default (/var/task/nextjs-store/.next/server/pages/api/success.js:57:115)
    at apiResolver (/var/task/nextjs-store/node_modules/next/dist/next-server/server/api-utils.js:8:7)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async Server.handleApiRequest (/var/task/nextjs-store/node_modules/next/dist/next-server/server/next-server.js:67:462)
    at async Object.fn (/var/task/nextjs-store/node_modules/next/dist/next-server/server/next-server.js:59:492)
    at async Router.execute (/var/task/nextjs-store/node_modules/next/dist/next-server/server/router.js:25:67)
    at async Server.run (/var/task/nextjs-store/node_modules/next/dist/next-server/server/next-server.js:69:1042)
    at async Server.handleRequest (/var/task/nextjs-store/node_modules/next/dist/next-server/server/next-server.js:34:504)
    at async Server.<anonymous> (/var/task/nextjs-store/___next_launcher.js:26:9)

```js
import eBayApi from "@hendt/ebay-api";
import prisma from "../../lib/prisma";

const eBay = new eBayApi({
  appId: "SimonEdg-StoreApp-SBX-acbb713f2-79a757fb",
  certId: "SBX-cbb713f2f5de-de52-421c-b6ef-f89f",
  sandbox: true,

  scope: [
    "https://api.ebay.com/oauth/api_scope",
    "https://api.ebay.com/oauth/api_scope/sell.fulfillment.readonly",
    "https://api.ebay.com/oauth/api_scope/sell.fulfillment",
  ],

  // optional parameters, should be omitted if not used
  siteId: eBayApi.SiteId.EBAY_UK, // required for traditional APIs, see https://developer.ebay.com/DevZone/merchandising/docs/Concepts/SiteIDToGlobalID.html
  //devId: "6ad92125-6505-40a5-b17d-58a0ad46e7db", // required for traditional trading API
  ruName: "Simon_Edge-SimonEdg-StoreA-rcrfx", // required for authorization code grant
  //authToken: "", // can be set to use traditional API without code grant
});

export default async (req, res) => {
  // Access the provided 'page' and 'limt' query parameters
  const code = req.query.code; // this is provided from eBay
  console.log(code);

  try {
    //const token = await eBay.OAuth2.getToken(code);
    const token = "bob";

    console.log("Prisma handler instance", prisma);

    const env_variable = await prisma.variable.upsert({
      where: {
        variable: "EBAY_TOKEN",
      },
      update: { value: token },
      create: {
        variable: "EBAY_TOKEN",
        value: token,
      },
    });

    if (env_variable) {
      console.log("New Token Stored in DB");
    } else console.log("Failed to store new Token");

    res.status(200);
    res.writeHead(302, {
      Location: "/orders",
      //add other headers here...
    });
    res.end();
  } catch (e) {
    console.error(e);
    res.status(400).end();
  }

  res.writeHead(302, {
    Location: "/orders",
    //add other headers here...
  });
  res.end();
};
```

**Reply by 2color** [reply]
Thanks for sharing @sunrunner4kr-simon. Could you maybe share a reproduction repository so that we can take a closer look at this?

**Reply by sunrunner4kr-simon** [reply]
ah yes, sorry. It's on Github, I can give you access: - 

https://github.com/sunrunner4kr-simon/StoreApp

**Reply by janpio** [reply]
Can you please open an issue @sunrunner4kr-simon as a bug report? Then it is much easier for us to label and investiagte this than here in this GH discussion. Just include the latest version of the error you are getting, a link to the reproduction and of course this discussion for context. Thanks.

**Reply by sunrunner4kr-simon** [reply]
done, thanks both

**Reply by sunrunner4kr-simon** [reply, marked as answer]
So, I had a play around, and think I found the problem. My prisma table field was a VARCHAR (String), however I was inadvertently trying to store upsert with a JSON object.
Now that I've changed to a JSON field it's working.

So I guess the only problem is that maybe the error wasn't helpul?
Although it was all my stupid fault.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
